streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamGraphNode.java

```diff
@@ -0,0 +1,106 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+
+abstract class StreamGraphNode<K, V> {
+
+    private StreamGraphNode<K, V> parentNode;
+    private final Collection<StreamGraphNode<K, V>> childNodes = new LinkedHashSet<>();
+    private final String nodeName;
+    private String parentNodeName;
+    private boolean repartitionRequired;
+    private boolean triggersRepartitioning;
+    private Integer id;
+    private StreamsTopologyGraphImpl streamsTopologyGraph;
+
+    StreamGraphNode(final String parentNodeName,
+                    final String nodeName,
+                    final boolean repartitionRequired) {
+        this.parentNodeName = parentNodeName;
+        this.nodeName = nodeName;
+        this.repartitionRequired = repartitionRequired;
+    }
+
+    StreamGraphNode<K, V> parentNode() {
+        return parentNode;
+    }
+
+    String parentNodeName() {
+        return parentNodeName;
+    }
+
+    void setParentNodeName(final String parentNodeName) {
+        this.parentNodeName = parentNodeName;
+    }
+
+    void setParentNode(final StreamGraphNode<K, V> parentNode) {
+        this.parentNode = parentNode;
+    }
+
+    Collection<StreamGraphNode<K, V>> children() {
+        return new LinkedHashSet<>(childNodes);
+    }
+
+    void addChildNode(final StreamGraphNode<K, V> node) {
+        this.childNodes.add(node);
+    }
+
+    String nodeName() {
+        return nodeName;
+    }
+
+    boolean isRepartitionRequired() {
+        return repartitionRequired;
+    }
+
+    void setRepartitionRequired(boolean repartitionRequired) {
+        this.repartitionRequired = repartitionRequired;
+    }
+
+    public boolean isTriggersRepartitioning() {
+        return triggersRepartitioning;
+    }
+
+    public void setTriggersRepartitioning(final boolean triggersRepartitioning) {
+        this.triggersRepartitioning = triggersRepartitioning;
+    }
+
+    void setId(final int id) {
+        this.id = id;
+    }
+
+    Integer id() {
+        return this.id;
+    }
+
+    public void setStreamsTopologyGraph(final StreamsTopologyGraphImpl streamsTopologyGraph) {
+        this.streamsTopologyGraph = streamsTopologyGraph;
+    }
+
+    StreamsTopologyGraphImpl streamsTopologyGraph() {
+        return streamsTopologyGraph;
+    }
+
+    abstract void writeToTopology(final InternalTopologyBuilder topologyBuilder);
```

Comment:
I added abstract method as each node should know how to write its details to the topology.  I don't consider this a leaking of implementation details as each node already contains all the parts needed to pass to the `InternalTopologyBuilder`.  
Additionally, each node has access to the `StreamsTopologyGraph` so each node can lookup the metadata created when the DSL created the graph.  So any optimization decisions can be made at this point as well.